### DISCUSSION TITLE
Is there a way to disable the extensions of a dependency that uses PyO3?

### DISCUSSION THREAD

**Discussion by simgt** [PyO3/pyo3#2798] [Questions] [answered]
Following my question on https://github.com/insight-platform/Similari/issues/87, I am looking for a way to disable as much as possible of a library's Python bindings to avoid relying on having Python available at build and run time.

Following the [features reference](https://pyo3.rs/latest/features.html) page of the doc, I have tried adding the following to the dependency's Cargo.toml:

```toml
pyo3 = { version = "0.16", default-features = false, features = ["abi3"] }
```

But the build still fails with:

```
error: failed to run custom build command for `pyo3-ffi v0.16.6`

Caused by:
  process didn't exit successfully: `...` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=PYO3_CROSS
  cargo:rerun-if-env-changed=PYO3_CROSS_LIB_DIR
  cargo:rerun-if-env-changed=PYO3_CROSS_PYTHON_VERSION
  cargo:rerun-if-env-changed=PYO3_CROSS_PYTHON_IMPLEMENTATION
  cargo:rerun-if-env-changed=PYO3_PRINT_CONFIG

  --- stderr
  error: the configured Python interpreter version (3.6) is lower than PyO3's minimum supported version (3.7)
```

Is there any built in way to achieve this?

**Comment by messense** [comment]
You can make python binding deps optional via cargo features.

```toml
[dependencies.pyo3]
version = "0.16"
features = ["extension-module"]
optional = true

[build-dependencies]
pyo3-build-config = { version = "0.16", optional = true }

[features]
python = ["pyo3", "pyo3-build-config"]
```

and `cfg`-gate your python binding api in code, for example

```rust
#[cfg(feature = "python")]
mod python {
    // your python binding code
}
```

And add `tool.maturin.features` to pyproject.toml to enable `python` feature when building with maturin

```toml
[tool.maturin]
features = ["python"]
```

**Reply by simgt** [reply]
Thanks a lot! That's what I was thinking of initially but I was hoping for a more automatic way to do it given that most things go through macros.

**Comment by adamreichold** [comment, marked as answer]
Additionally, if the above to invasive a change for your dependency, maybe you could try `features = ["abi-py37", "generate-abi3-import-lib"]` which means we link against a prebuilt library and should work without a Python interpreter being present.

**Reply by simgt** [reply]
That doesn't seem to work, sadly. I'm getting the same error with `generate-import-lib`. Maybe I missed something?

**Reply by adamreichold** [reply]
It is possible that we wired this up only for Windows cross builds, but I am not sure at the moment and pressed for time to check it.

**Reply by adamreichold** [reply]
Ok, so now I had time to read up on https://pyo3.rs/latest/building_and_distribution.html#building-abi3-extensions-without-a-python-interpreter and on Unix-like targets this should work out of the box, but for Windows, the `generate-abi3-import-lib` feature is required.

In any case, as the docs say you might need to set the `PYO3_NO_PYTHON` environment variable to suppress the automatic detection of the installed Python interpreter (which seems to be too old in your case).

**Reply by simgt** [reply]
Ha perfect, that solves most of my issue. I completely missed this section when browsing the manual. Thanks a lot for following up!

The resulting program could be smaller, but adding an optional feature to the dependency is a much bigger task so that'll be good enough for now.
